client/document/index.jsx

```diff
@@ -125,32 +130,14 @@ class Document extends React.Component {
 						</div>
 					) }
 					{ badge && (
-						<div className="environment-badge">
-							{ preferencesHelper && <div className="environment is-prefs" /> }
-							{ abTestHelper && <div className="environment is-tests" /> }
-							{ branchName &&
-								branchName !== 'master' && (
-									<span className="environment branch-name" title={ 'Commit ' + commitChecksum }>
-										{ branchName }
-									</span>
-								) }
-							{ devDocs && (
-								<span className="environment is-docs">
-									<a href={ devDocsURL } title="DevDocs">
-										docs
-									</a>
-								</span>
+						<EnvironmentBadge badge={ badge } feedbackURL={ feedbackURL }>
+							{ preferencesHelper && <PreferencesHelper /> }
+							{ abTestHelper && <TestHelper /> }
+							{ branchName && (
+								<Branch branchName={ branchName } commitChecksum={ commitChecksum } />
 							) }
-							<span className={ `environment is-${ badge } is-env` }>{ badge }</span>
-							<a
-								className="bug-report"
-								href={ feedbackURL }
-								title="Report an issue"
-								target="_blank"
-							>
-								<Gridicon icon="bug" size={ 18 } />
-							</a>
-						</div>
+							{ devDocs && <DevDocsLink url={ devDocsURL } /> }
+						</EnvironmentBadge>
```

Comment:
What about passing `preferencesHelper`, `abTestHelper`, `branchName`, and `devDocs` as props instead and move the conditionals whether to display those child components to the `EnvironmentBadge` component?

Reply:
(In that case, we could probably ditch `devDocs` and just check for presence of `devDocsUrl`.)

Reply:
@ockham my plan was to have a more flexible component which doesn't contain too much logic but rather have the surrounding component (in this case our document) decide what "features" of the badge component to use. Imo this makes it easier to create additional or change child components and/or reuse the badge component in places where not all of the features are necessary which we use here.

Reply:
Well, there might be a tradeoff in terms of reusability, but I think I'd prefer things to be a bit more contained in this case...

Reply:
Can you elaborate on what you mean by "more contained" and why you think it's preferable? I still think it will be easier to add, change or remove specific child components if they are more independent from the parent component like `li` is to `ul`.